src/Persistence_matrix/include/gudhi/matrix.h

```diff
@@ -0,0 +1,2068 @@
+   * (after creating the matrix with the @ref Matrix(int numberOfColumns, characteristic_type characteristic)
+   * constructor preferably). If the persistence barcode has to be computed from this matrix, the simplices are also
+   * assumed to be ordered by appearance order in the filtration. Also, depending of the options, the matrix is
+   * eventually reduced on the fly or converted into a chain complex base, so the new matrix is not always identical to
+   * the old one.
+   * @param characteristic Characteristic of the coefficient field. Has to be specified if
+   * @ref PersistenceMatrixOptions::is_z2 is false. Default value is 11. Ignored if
+   * @ref PersistenceMatrixOptions::is_z2 is true.
+   */
+  template <class Container_type = boundary_type>
+  Matrix(const std::vector<Container_type>& columns, characteristic_type characteristic = 11);
+  /**
+   * @brief Constructs a new empty matrix and reserves space for the given number of columns.
+   *
+   * @param numberOfColumns Number of columns to reserve space for.
+   * @param characteristic Characteristic of the coefficient field. If not specified and
+   * @ref PersistenceMatrixOptions::is_z2 is false, the characteristic has to be set later with the use of
+   * @ref set_characteristic before calling for the first time a method needing it. Ignored if
+   * @ref PersistenceMatrixOptions::is_z2 is true.
+   */
+  Matrix(int numberOfColumns, characteristic_type characteristic = 0);
+  /**
+   * @brief Constructs a new empty matrix with the given comparator functions. Only available when those comparators
+   * are necessary.
+   *
+   * That is, when **all** following options have following values:
+   *   - @ref PersistenceMatrixOptions::is_of_boundary_type = false
+   *   - @ref PersistenceMatrixOptions::has_vine_update = true
+   *   - @ref PersistenceMatrixOptions::has_column_pairings = false
+   *
+   * Those comparators are necesseray to distinguish cases in a vine update. When the matrix is of
+   * @ref boundarymatrix "boundary type" or if the column pairing is activated (i.e., the barcode is stored),
+   * the comparators can be easily deduced without overhead. If neither are true, we assume that one has additional
+   * information outside of the matrix about the barcode to provide a better suited comparator adapted to the situation
+   * (as in the implementation of the Zigzag algorithm @cite zigzag for example.)
+   *
+   * @param birthComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller birth than the bar associated to the second one.
+   * @param deathComparator Method taking two @ref PosIdx indices as parameter and returns true if and only if the first
+   * face is associated to a bar with strictly smaller death than the bar associated to the second one.
+   */
+  Matrix(const std::function<bool(pos_index,pos_index)>& birthComparator, 
```

Comment:
I remember you needed `std::function` because the comparator is not always a plain function, but did you have a chance to check if it had noticable overhead?

Reply:
No I couldn't check, because it just didn't worked otherwise...